sklearn/metrics/pairwise.py

```diff
@@ -219,34 +255,89 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     --------
     paired_distances : distances betweens pairs of elements of X and Y.
     """
-    X, Y = check_pairwise_arrays(X, Y)
 
-    if X_norm_squared is not None:
-        XX = check_array(X_norm_squared)
-        if XX.shape == (1, X.shape[0]):
-            XX = XX.T
-        elif XX.shape != (X.shape[0], 1):
-            raise ValueError(
-                "Incompatible dimensions for X and X_norm_squared")
-    else:
-        XX = row_norms(X, squared=True)[:, np.newaxis]
-
-    if X is Y:  # shortcut in the common case euclidean_distances(X, X)
-        YY = XX.T
-    elif Y_norm_squared is not None:
-        YY = np.atleast_2d(Y_norm_squared)
+    # NOTE: force_all_finite=False allows not only NaN but also inf/-inf
+    X, Y = check_pairwise_arrays(X, Y,
+                                 force_all_finite=kill_missing, copy=copy)
+    if kill_missing is False and \
+            (np.any(np.isinf(X.data)) or
+                (Y is not None and np.any(np.isinf(Y.data)))):
+        raise ValueError(
+            "+/- Infinite values are not allowed.")
+
+    if kill_missing:
+        if X_norm_squared is not None:
+            XX = check_array(X_norm_squared)
+            if XX.shape == (1, X.shape[0]):
+                XX = XX.T
+            elif XX.shape != (X.shape[0], 1):
+                raise ValueError(
+                    "Incompatible dimensions for X and X_norm_squared")
+        else:
+            XX = row_norms(X, squared=True)[:, np.newaxis]
```

Comment:
Maybe I've just misunderstood what's going on here. Can you please make a separate function rather than having a `kill_missing` setting?